| Nama                                | Kelas | NIM        |
| ----------------------------------- | ----- | ---------- |
| Ulfi Mustatiq Abidatul Izza         | TI-3H | 2141720052 |

# Tugas

1. Buatlah model klasifikasi dengan menggunakan SVM untuk data suara, voice.csv. 

2. Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,

a. Menggunakan data spam.csv

b. Fitur CountVectorizer dengan mengaktifkan stop_words

c. Evaluasi hasilnya


3. Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,

a. Menggunakan data spam.csv

b. Fitur TF-IDF dengan mengaktifkan stop_words

c. Evaluasi hasilnya dan bandingkan dengan hasil pada Tugas no 2.


Berikan kesimpulan fitur mana yang terbaik pada kasus data spam.csv

**Jawab**

## 1. Model Klasifikasi menggunakan SVM pada Data Suara

In [1]:
# import library
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats
import seaborn as sns
import pandas as pd
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# baca data dari file CSV
data = pd.read_csv('data/voice.csv')

# melihat beberapa data awal
data.head()

meanfreq        sd    median       Q25       Q75       IQR       skew  \
0  0.059781  0.064241  0.032027  0.015071  0.090193  0.075122  12.863462   
1  0.066009  0.067310  0.040229  0.019414  0.092666  0.073252  22.423285   
2  0.077316  0.083829  0.036718  0.008701  0.131908  0.123207  30.757155   
3  0.151228  0.072111  0.158011  0.096582  0.207955  0.111374   1.232831   
4  0.135120  0.079146  0.124656  0.078720  0.206045  0.127325   1.101174   

          kurt    sp.ent       sfm  ...  centroid   meanfun    minfun  \
0   274.402906  0.893369  0.491918  ...  0.059781  0.084279  0.015702   
1   634.613855  0.892193  0.513724  ...  0.066009  0.107937  0.015826   
2  1024.927705  0.846389  0.478905  ...  0.077316  0.098706  0.015656   
3     4.177296  0.963322  0.727232  ...  0.151228  0.088965  0.017798   
4     4.333713  0.971955  0.783568  ...  0.135120  0.106398  0.016931   

     maxfun   meandom    mindom    maxdom   dfrange   modindx  label  
0  0.275862  0.007812  0.007812  0.007812  0.000000  0.000000   male  
1  0.250000  0.009014  0.007812  0.054688  0.046875  0.052632   male  
2  0.271186  0.007990  0.007812  0.015625  0.007812  0.046512   male  
3  0.250000  0.201497  0.007812  0.562500  0.554688  0.247119   male  
4  0.266667  0.712812  0.007812  5.484375  5.476562  0.208274   male  

[5 rows x 21 columns]

##### Encoding dan Membagi Data Menjadi Atribut dan Target

In [2]:
# Mengganti label kategori dengan angka (misal: 'male' menjadi 0, 'female' menjadi 1)
label_encoder = LabelEncoder()
data['label'] = label_encoder.fit_transform(data['label'])

# Memisahkan atribut (fitur) dan target
X = data.drop('label', axis=1)
y = data['label']

##### Membagi Data menjadi Data Pelatihan dan Data Pengujian

In [3]:
# Membagi data menjadi data pelatihan (70%) dan data pengujian (30%)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

##### Membuat Model SVM

In [4]:
# Membuat model SVM dengan kernel linear
svm_model = SVC(kernel='linear', random_state=42)

##### Melatih Model SVM

In [5]:
# Melatih model SVM
svm_model.fit(X_train, y_train)

SVC(kernel='linear', random_state=42)

##### Evaluasi Model SVM

In [6]:
# Memprediksi label untuk data pengujian
y_pred = svm_model.predict(X_test)

# Mengukur akurasi model
accuracy = accuracy_score(y_test, y_pred)
print("Akurasi Model SVM:", accuracy)

# Menampilkan laporan klasifikasi
print("\nLaporan Klasifikasi:")
print(classification_report(y_test, y_pred))

# Menampilkan matriks konfusi
print("\nMatriks Konfusi:")
print(confusion_matrix(y_test, y_pred))

Akurasi Model SVM: 0.9263932702418507

Laporan Klasifikasi:
              precision    recall  f1-score   support

           0       0.98      0.86      0.92       452
           1       0.89      0.98      0.93       499

    accuracy                           0.93       951
   macro avg       0.93      0.92      0.93       951
weighted avg       0.93      0.93      0.93       951


Matriks Konfusi:
[[390  62]
 [  8 491]]


##### Akurasi Model SVM Terhadap Label Sebenarnya dan Hasil Prediksi

In [8]:
# Hitung akurasi
accuracy = accuracy_score(y_test, y_pred)

# Cetak hasil akurasi
print(f'Skor Akurasi : {int(accuracy * 100)}%')

Skor Akurasi : 92%


## 2. Model klasfikasi Multinomial Naive Bayes

##### a. Menggunakan data spam.csv

In [9]:
# baca data dari file CSV
data = pd.read_csv('data/spam.csv')

# melihat beberapa data awal
data.head()

v1                                                 v2 Unnamed: 2  \
0   ham  Go until jurong point, crazy.. Available only ...        NaN   
1   ham                      Ok lar... Joking wif u oni...        NaN   
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3   ham  U dun say so early hor... U c already then say...        NaN   
4   ham  Nah I don't think he goes to usf, he lives aro...        NaN   

  Unnamed: 3 Unnamed: 4  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4        NaN        NaN

##### Rename pada Nama Kolom v1 dan v2

In [12]:
# Drop 3 kolom terakhir dengan fungsi iloc
data = data.drop(data.iloc[:,2:], axis=1)

# Data untuk rename kolom v1 dan v2
new_cols = {
    'v1': 'Labels',
    'v2': 'SMS'
}

# Rename nama kolom v1 dan v2
data = data.rename(columns=new_cols)

# cek data
data.head()

Labels                                                SMS
0    ham  Go until jurong point, crazy.. Available only ...
1    ham                      Ok lar... Joking wif u oni...
2   spam  Free entry in 2 a wkly comp to win FA Cup fina...
3    ham  U dun say so early hor... U c already then say...
4    ham  Nah I don't think he goes to usf, he lives aro...

##### Inspeksi data

In [13]:
# Cek Jumlah Data Per Kelas
print(data['Labels'].value_counts())
print('\n')

# Cek Kelengkapan Data
print(data.info())
print('\n')

# Cek Statistik Deskriptif
print(data.describe())

ham     4825
spam     747
Name: Labels, dtype: int64


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5572 entries, 0 to 5571
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Labels  5572 non-null   object
 1   SMS     5572 non-null   object
dtypes: object(2)
memory usage: 87.2+ KB
None


       Labels                     SMS
count    5572                    5572
unique      2                    5169
top       ham  Sorry, I'll call later
freq     4825                      30


##### Encoding dan Memisahkan Atribut dan Target

In [14]:
# Mengganti label kategori dengan angka
data['Labels'] = data['Labels'].map({'ham': 0, 'spam': 1})

# Memisahkan atribut (fitur) dan target
X = data['SMS'].values
y = data['Labels'].values

##### b. Fitur CountVectorizer dengan mengaktifkan stop_words

##### Membagi Data menjadi Data Pelatihan dan Data Pengujian

In [15]:
# Membagi data menjadi data pelatihan (70%) dan data pengujian (30%)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

##### Membuat Model Multinomial Naive Bayes dengan CountVectorizer

In [16]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.pipeline import Pipeline

# Membuat pipeline dengan CountVectorizer dan Multinomial Naive Bayes
pipeline = Pipeline([
    ('vectorizer', CountVectorizer(stop_words='english')),
    ('classifier', MultinomialNB())
])

# Melatih model
pipeline.fit(X_train, y_train)


Pipeline(steps=[('vectorizer', CountVectorizer(stop_words='english')),
                ('classifier', MultinomialNB())])

##### c. Evaluasi Hasilnya

In [17]:
# Memprediksi label untuk data pengujian
y_pred = pipeline.predict(X_test)

# Mengukur akurasi model
accuracy = accuracy_score(y_test, y_pred)
print("Akurasi Model Multinomial Naive Bayes:", accuracy)

# Menampilkan laporan klasifikasi
print("\nLaporan Klasifikasi:")
print(classification_report(y_test, y_pred))

# Menampilkan matriks konfusi
print("\nMatriks Konfusi:")
print(confusion_matrix(y_test, y_pred))

Akurasi Model Multinomial Naive Bayes: 0.9826555023923444

Laporan Klasifikasi:
              precision    recall  f1-score   support

           0       0.99      0.99      0.99      1453
           1       0.95      0.92      0.93       219

    accuracy                           0.98      1672
   macro avg       0.97      0.96      0.96      1672
weighted avg       0.98      0.98      0.98      1672


Matriks Konfusi:
[[1442   11]
 [  18  201]]


In [26]:
# Inisiasi MultinomialNB
mnb = MultinomialNB()

# Memprediksi label untuk data pelatihan
y_train_pred = pipeline.predict(X_train)

# Mengukur akurasi model pada data pelatihan
train_accuracy = accuracy_score(y_train, y_train_pred)
print("Akurasi Model Multinomial Naive Bayes pada Data Pelatihan:", train_accuracy)

# Memprediksi label untuk data pengujian
y_test_pred = pipeline.predict(X_test)

# Mengukur akurasi model pada data pengujian
test_accuracy = accuracy_score(y_test, y_test_pred)
print("Akurasi Model Multinomial Naive Bayes pada Data Pengujian:", test_accuracy)

# Evaluasi akurasi data training
acc_train = accuracy_score(y_train, y_train_pred)

# Evaluasi akurasi data test
acc_test = accuracy_score(y_test, y_test_pred)

# Print hasil evaluasi
print(f'Hasil persentase akurasi data train: {acc_train * 100:2f}%')
print(f'Hasil persentase akurasi data test: {acc_test * 100:2f}%')

Akurasi Model Multinomial Naive Bayes pada Data Pelatihan: 0.9953846153846154
Akurasi Model Multinomial Naive Bayes pada Data Pengujian: 0.9826555023923444
Hasil persentase akurasi data train: 99.538462%
Hasil persentase akurasi data test: 98.265550%


## 3. Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,

##### a. Menggunakan data spam.csv

In [27]:
# baca data dari file CSV
data = pd.read_csv('data/spam.csv')

# melihat beberapa data awal
data.head()

v1                                                 v2 Unnamed: 2  \
0   ham  Go until jurong point, crazy.. Available only ...        NaN   
1   ham                      Ok lar... Joking wif u oni...        NaN   
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3   ham  U dun say so early hor... U c already then say...        NaN   
4   ham  Nah I don't think he goes to usf, he lives aro...        NaN   

  Unnamed: 3 Unnamed: 4  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4        NaN        NaN

##### Hapus 3 Kolom Terakhir dan Rename Nama Kolom v1 dan v2

In [28]:
# Drop 3 kolom terakhir dengan fungsi iloc
data = data.drop(data.iloc[:,2:], axis=1)

# Data untuk rename kolom v1 dan v2
new_cols = {
    'v1': 'Labels',
    'v2': 'SMS'
}

# Rename nama kolom v1 dan v2
data = data.rename(columns=new_cols)

# cek data
data.head()

Labels                                                SMS
0    ham  Go until jurong point, crazy.. Available only ...
1    ham                      Ok lar... Joking wif u oni...
2   spam  Free entry in 2 a wkly comp to win FA Cup fina...
3    ham  U dun say so early hor... U c already then say...
4    ham  Nah I don't think he goes to usf, he lives aro...

##### Encoding dan Memisahkan Aribut dan Target

In [29]:
# Mengganti label kategori dengan angka
data['Labels'] = data['Labels'].map({'ham': 0, 'spam': 1})

# Memisahkan atribut (fitur) dan target
X = data['SMS'].values
y = data['Labels'].values

##### b. Fitur TF-IDF dengan mengaktifkan stop_words

##### Membagi Data menjadi Data Pelatihan dan Data Pengujian

In [30]:
# Membagi data menjadi data pelatihan (70%) dan data pengujian (30%)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

##### Membuat Model Multinomial Naive Bayes dengan TF-IDF Vectorizer

In [31]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.pipeline import Pipeline

# Membuat pipeline dengan TfidfVectorizer dan Multinomial Naive Bayes
pipeline = Pipeline([
    ('vectorizer', TfidfVectorizer(stop_words='english')),
    ('classifier', MultinomialNB())
])

# Melatih model
pipeline.fit(X_train, y_train)

Pipeline(steps=[('vectorizer', TfidfVectorizer(stop_words='english')),
                ('classifier', MultinomialNB())])

##### c. Evaluasi hasilnya dan bandingkan dengan hasil pada Tugas no 2.

##### Evaluasi Model dengan TF-IDF

In [32]:
# Memprediksi label untuk data pengujian
y_pred = pipeline.predict(X_test)

# Mengukur akurasi model dengan TF-IDF
accuracy_tfidf = accuracy_score(y_test, y_pred)
print("Akurasi Model Multinomial Naive Bayes dengan TF-IDF:", accuracy_tfidf)

# Menampilkan laporan klasifikasi dengan TF-IDF
print("\nLaporan Klasifikasi dengan TF-IDF:")
print(classification_report(y_test, y_pred))

# Menampilkan matriks konfusi dengan TF-IDF
print("\nMatriks Konfusi dengan TF-IDF:")
print(confusion_matrix(y_test, y_pred))

Akurasi Model Multinomial Naive Bayes dengan TF-IDF: 0.9677033492822966

Laporan Klasifikasi dengan TF-IDF:
              precision    recall  f1-score   support

           0       0.96      1.00      0.98      1453
           1       1.00      0.75      0.86       219

    accuracy                           0.97      1672
   macro avg       0.98      0.88      0.92      1672
weighted avg       0.97      0.97      0.97      1672


Matriks Konfusi dengan TF-IDF:
[[1453    0]
 [  54  165]]


In [35]:
# Memprediksi label untuk data pelatihan
y_train_pred = pipeline.predict(X_train)

# Mengukur akurasi model dengan TF-IDF pada data pelatihan
accuracy_train_tfidf = accuracy_score(y_train, y_train_pred)
print("Akurasi Model Multinomial Naive Bayes dengan TF-IDF pada Data Pelatihan:", accuracy_train_tfidf)

# Memprediksi label untuk data pengujian
y_test_pred = pipeline.predict(X_test)

# Mengukur akurasi model dengan TF-IDF pada data pengujian
accuracy_test_tfidf = accuracy_score(y_test, y_test_pred)
print("Akurasi Model Multinomial Naive Bayes dengan TF-IDF pada Data Pengujian:", accuracy_test_tfidf)

# Evaluasi akurasi data training
acc_train = accuracy_score(y_train, y_train_pred)

# Evaluasi akurasi data test
acc_test = accuracy_score(y_test, y_test_pred)

# Print hasil evaluasi
print(f'Hasil persentase akurasi data train: {acc_train * 100:f}%')
print(f'Hasil persentase akurasi data test: {acc_test * 100:f}%')

Akurasi Model Multinomial Naive Bayes dengan TF-IDF pada Data Pelatihan: 0.9825641025641025
Akurasi Model Multinomial Naive Bayes dengan TF-IDF pada Data Pengujian: 0.9677033492822966
Hasil persentase akurasi data train: 98.256410%
Hasil persentase akurasi data test: 96.770335%


##### d. Berikan kesimpulan fitur mana yang terbaik pada kasus data spam.csv

Berdasarkan dari hasil evaluasi model klasfikasi Multinomial Naive Bayes dengan menggunakan Fitur CountVectorizer dan Fitur TF-IDF adalah :

**Fitur CountVectorizer**
- Hasil persentase akurasi data train: 99.538462%
- Hasil persentase akurasi data test: 98.265550%

**Fitur TF-IDF**
- Hasil persentase akurasi data train: 98.256410%
- Hasil persentase akurasi data test: 96.770335%

Dalam kasus ini, fitur CountVectorizer memberikan hasil akurasi yang lebih tinggi baik pada data pelatihan maupun data pengujian dibandingkan dengan fitur TF-IDF. Oleh karena itu, berdasarkan hasil akurasi yang diperoleh, dapat disimpulkan bahwa fitur CountVectorizer lebih baik dalam kasus data "spam.csv." Fitur ini tampaknya lebih efektif dalam mengklasifikasikan data spam dan ham dalam dataset tersebut.

Namun, penting untuk diingat bahwa pemilihan fitur tergantung pada konteks dan karakteristik data tertentu. Pada kasus lain atau dengan dataset yang berbeda, hasilnya mungkin berbeda.